File: app/src/main/java/com/cjstudio/sosestrada/CadastroMotoristaActivity.java
package com.cjstudio.sosestrada;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;

public class CadastroMotoristaActivity extends AppCompatActivity {

    private TextInputEditText edtNome, edtTelefone, edtEmail, edtSenha, edtVeiculo, edtPlaca, edtCor;
    private MaterialButton btnCadastrar;
    private TextView btnVoltar;
    private FirebaseAuth mAuth;
    private FirebaseFirestore db;
    private boolean isEditando = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cadastro_motorista);

        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();

        edtNome = findViewById(R.id.edtNome);
        edtTelefone = findViewById(R.id.edtTelefone);
        edtEmail = findViewById(R.id.edtEmail);
        edtSenha = findViewById(R.id.edtSenha);
        edtVeiculo = findViewById(R.id.edtVeiculo);
        edtPlaca = findViewById(R.id.edtPlaca);
        edtCor = findViewById(R.id.edtCor);
        btnCadastrar = findViewById(R.id.btnCadastrar);
        btnVoltar = findViewById(R.id.btnVoltar);

        Intent intent = getIntent();
        if (intent.getBooleanExtra("motorista", false)) {
            isEditando = true;
            edtNome.setText(intent.getStringExtra("motorista_nome"));
            edtTelefone.setText(intent.getStringExtra("motorista_telefone"));
            edtEmail.setText(intent.getStringExtra("motorista_email"));
            edtVeiculo.setText(intent.getStringExtra("motorista_veiculo"));
            edtPlaca.setText(intent.getStringExtra("motorista_placa"));
            edtCor.setText(intent.getStringExtra("motorista_cor"));
            btnCadastrar.setText("ATUALIZAR CADASTRO");
        } else {
            btnCadastrar.setText("CADASTRAR");
        }

        btnVoltar.setOnClickListener(v -> finish());

        btnCadastrar.setOnClickListener(v -> realizarCadastro());
    }

    private void realizarCadastro() {
        String nome = edtNome.getText().toString().trim();
        String telefone = edtTelefone.getText().toString().trim();
        String email = edtEmail.getText().toString().trim();
        String senha = edtSenha.getText().toString().trim();
        String veiculo = edtVeiculo.getText().toString().trim();
        String placa = edtPlaca.getText().toString().trim().toUpperCase();
        String cor = edtCor.getText().toString().trim();

        // Validações...
        if (TextUtils.isEmpty(nome)) {
            edtNome.setError("Nome obrigatório");
            edtNome.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(telefone)) {
            edtTelefone.setError("Telefone obrigatório");
            edtTelefone.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(email)) {
            edtEmail.setError("E-mail obrigatório");
            edtEmail.requestFocus();
            return;
        }
        if (!android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            edtEmail.setError("E-mail inválido");
            edtEmail.requestFocus();
            return;
        }
        if (!isEditando && TextUtils.isEmpty(senha)) {
            edtSenha.setError("Senha obrigatória para novo cadastro");
            edtSenha.requestFocus();
            return;
        }
        if (!isEditando && senha.length() < 6) {
            edtSenha.setError("Senha deve ter pelo menos 6 caracteres");
            edtSenha.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(veiculo)) {
            edtVeiculo.setError("Modelo do veículo obrigatório");
            edtVeiculo.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(placa)) {
            edtPlaca.setError("Placa obrigatória");
            edtPlaca.requestFocus();
            return;
        }
        if (TextUtils.isEmpty(cor)) {
            edtCor.setError("Cor do veículo obrigatória");
            edtCor.requestFocus();
            return;
        }

        // Caso seja edição, o usuário já está logado
        if (isEditando) {
            atualizarDadosNoFirestore();
            return;
        }

        // NOVO CADASTRO: criar usuário no Firebase Auth
        mAuth.createUserWithEmailAndPassword(email, senha)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        FirebaseUser user = mAuth.getCurrentUser();
                        if (user != null) {
                            // Salva os dados no Firestore com o UID do usuário
                            salvarDadosMotorista(user.getUid(), nome, telefone, email, veiculo, placa, cor);
                        }
                    } else {
                        Toast.makeText(this, "Erro ao criar conta: " + task.getException().getMessage(), Toast.LENGTH_LONG).show();
                    }
                });
    }

    private void salvarDadosMotorista(String uid, String nome, String telefone, String email, String veiculo, String placa, String cor) {
        Map<String, Object> motorista = new HashMap<>();
        motorista.put("nome", nome);
        motorista.put("telefone", telefone);
        motorista.put("email", email);
        motorista.put("veiculo", veiculo);
        motorista.put("placa", placa);
        motorista.put("cor", cor);
        motorista.put("uid", uid);

        db.collection("motoristas").document(uid)
                .set(motorista)
                .addOnSuccessListener(aVoid -> {
                    Toast.makeText(this, "✅ Cadastro realizado!", Toast.LENGTH_LONG).show();
                    finish();
                })
                .addOnFailureListener(e -> {
                    Toast.makeText(this, "Erro ao salvar dados: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                });
    }

    private void atualizarDadosNoFirestore() {
        FirebaseUser user = mAuth.getCurrentUser();
        if (user == null) {
            Toast.makeText(this, "Usuário não autenticado", Toast.LENGTH_SHORT).show();
            return;
        }
        String uid = user.getUid();
        String nome = edtNome.getText().toString().trim();
        String telefone = edtTelefone.getText().toString().trim();
        String email = edtEmail.getText().toString().trim();
        String veiculo = edtVeiculo.getText().toString().trim();
        String placa = edtPlaca.getText().toString().trim().toUpperCase();
        String cor = edtCor.getText().toString().trim();

        Map<String, Object> motorista = new HashMap<>();
        motorista.put("nome", nome);
        motorista.put("telefone", telefone);
        motorista.put("email", email);
        motorista.put("veiculo", veiculo);
        motorista.put("placa", placa);
        motorista.put("cor", cor);
        motorista.put("uid", uid);

        db.collection("motoristas").document(uid)
                .set(motorista)
                .addOnSuccessListener(aVoid -> {
                    Toast.makeText(this, "✅ Dados atualizados!", Toast.LENGTH_LONG).show();
                    finish();
                })
                .addOnFailureListener(e -> {
                    Toast.makeText(this, "Erro ao atualizar: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                });
    }
}

File: app/src/main/java/com/cjstudio/sosestrada/CadastroMotoristaActivity.kt
package com.cjstudio.sosestrada

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Patterns
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.cjstudio.sosestrada.databinding.ActivityCadastroMotoristaBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

// Cadastro novo (cria a conta + motoristas/{uid}) ou edição do cadastro do
// motorista logado (aberta por intentEdicao, carrega os dados do Firestore).
@AndroidEntryPoint
class CadastroMotoristaActivity : AppCompatActivity() {

    @Inject
    lateinit var authRepository: IAuthRepository

    @Inject
    lateinit var motoristaRepository: IMotoristaRepository

    private lateinit var binding: ActivityCadastroMotoristaBinding
    private var editando = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCadastroMotoristaBinding.inflate(layoutInflater)
        setContentView(binding.root)

        editando = intent.getBooleanExtra(EXTRA_EDITANDO, false)
        binding.btnCadastrar.text = if (editando) "ATUALIZAR CADASTRO" else "CADASTRAR"
        if (editando) carregarCadastro()

        binding.btnVoltar.setOnClickListener { finish() }
        binding.btnCadastrar.setOnClickListener { realizarCadastro() }
    }

    private fun carregarCadastro() {
        lifecycleScope.launch {
            motoristaRepository.buscarMeuCadastro()
                .onSuccess { motorista ->
                    if (motorista == null) return@onSuccess
                    binding.edtNome.setText(motorista.nome)
                    binding.edtTelefone.setText(motorista.telefone)
                    binding.edtEmail.setText(motorista.email)
                    binding.edtVeiculo.setText(motorista.veiculo)
                    binding.edtPlaca.setText(motorista.placa)
                    binding.edtCor.setText(motorista.cor)
                }
                .onFailure { e ->
                    Toast.makeText(this@CadastroMotoristaActivity, "Erro ao buscar dados: ${e.message}", Toast.LENGTH_SHORT).show()
                }
        }
    }

    private fun realizarCadastro() {
        val nome = binding.edtNome.textoLimpo()
        val telefone = binding.edtTelefone.textoLimpo()
        val email = binding.edtEmail.textoLimpo()
        val senha = binding.edtSenha.textoLimpo()
        val veiculo = binding.edtVeiculo.textoLimpo()
        val placa = binding.edtPlaca.textoLimpo().uppercase()
        val cor = binding.edtCor.textoLimpo()

        when {
            nome.isEmpty() -> return binding.edtNome.erro("Nome obrigatório")
            telefone.isEmpty() -> return binding.edtTelefone.erro("Telefone obrigatório")
            email.isEmpty() -> return binding.edtEmail.erro("E-mail obrigatório")
            !Patterns.EMAIL_ADDRESS.matcher(email).matches() -> return binding.edtEmail.erro("E-mail inválido")
            !editando && senha.isEmpty() -> return binding.edtSenha.erro("Senha obrigatória para novo cadastro")
            !editando && senha.length < 6 -> return binding.edtSenha.erro("Senha deve ter pelo menos 6 caracteres")
            veiculo.isEmpty() -> return binding.edtVeiculo.erro("Modelo do veículo obrigatório")
            placa.isEmpty() -> return binding.edtPlaca.erro("Placa obrigatória")
            cor.isEmpty() -> return binding.edtCor.erro("Cor do veículo obrigatória")
        }

        val motorista = Motorista(nome = nome, telefone = telefone, email = email, veiculo = veiculo, placa = placa, cor = cor)
        binding.btnCadastrar.isEnabled = false
        lifecycleScope.launch {
            // Edição: o motorista já está logado. Cadastro novo: cria a conta antes.
            if (!editando) {
                val conta = authRepository.criarConta(email, senha)
                if (conta.isFailure) {
                    binding.btnCadastrar.isEnabled = true
                    Toast.makeText(this@CadastroMotoristaActivity, "Erro ao criar conta: ${conta.exceptionOrNull()?.message}", Toast.LENGTH_LONG).show()
                    return@launch
                }
            }
            motoristaRepository.salvarMeuCadastro(motorista)
                .onSuccess {
                    val mensagem = if (editando) "✅ Dados atualizados!" else "✅ Cadastro realizado!"
                    Toast.makeText(this@CadastroMotoristaActivity, mensagem, Toast.LENGTH_LONG).show()
                    finish()
                }
                .onFailure { e ->
                    binding.btnCadastrar.isEnabled = true
                    val prefixo = if (editando) "Erro ao atualizar" else "Erro ao salvar dados"
                    Toast.makeText(this@CadastroMotoristaActivity, "$prefixo: ${e.message}", Toast.LENGTH_SHORT).show()
                }
        }
    }

    private fun EditText.textoLimpo() = text.toString().trim()

    private fun EditText.erro(mensagem: String) {
        error = mensagem
        requestFocus()
    }

    companion object {
        private const val EXTRA_EDITANDO = "motorista"

        fun intentEdicao(context: Context): Intent =
            Intent(context, CadastroMotoristaActivity::class.java).putExtra(EXTRA_EDITANDO, true)
    }
}
